### ISSUE TITLE
Wrong distribution sampled by torch.multinomial on CUDA

### ISSUE THREAD

**Issue by LeviViana** [pytorch/pytorch#22086]
## 🐛 Bug

`torch.multinomial` is not sampling the correct distribution for CUDA tensors.

## To Reproduce

Here I compare the distributions sampled by the `torch.multinomial` on CUDA vs CPU tensors. The sampling is working correctly on CPU.

I've noticed this bug only for **sampling WITHOUT replacement**.

```python
import torch
import torch.nn.functional as F
import scipy.stats
import numpy as np

m = 10
n = 10000
replace = False

#################################################
# Comparing Multinomial CPU vs Multinomial CUDA #
#################################################

def check_for_k(k):

	multinomial_cuda_samples = []
	multinomial_cpu_samples = []

	weights = torch.rand(m)

	for _ in range(n):
		multinomial_cuda_samples += torch.multinomial(
							  weights.cuda(),
							  k,
							  replace
							).cpu().numpy().tolist()

		multinomial_cpu_samples += torch.multinomial(
							  weights.cpu(),
							  k,
							  replace
							).cpu().numpy().tolist()


	_, multinomial_cuda_dist = np.unique(multinomial_cuda_samples, return_counts=True)
	_, multinomial_cpu_dist = np.unique(multinomial_cpu_samples, return_counts=True)

	_, p = scipy.stats.chisquare(multinomial_cuda_dist, multinomial_cpu_dist)
	print("Chi-Squared Test p-value {:.3f}".format(p))

	multinomial_cuda_dist = torch.Tensor(multinomial_cuda_dist) / (n * k)
	multinomial_cpu_dist = torch.Tensor(multinomial_cpu_dist) / (n * k)

	print("KL Divergence {:.3f}".format(F.kl_div(
						multinomial_cpu_dist.log(), 
						multinomial_cuda_dist, 
						reduction='sum'
						)
					)
	)


print("CHECKING for k = 1")
check_for_k(1)
print("CHECKING for k = 5")
check_for_k(5)
```

## Expected behavior

Sample from a multinomial distribution on CUDA and CPU tensors. 

## Environment

```bash
PyTorch version: 1.2.0.dev20190621
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Amazon Linux 2
GCC version: (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
CMake version: version 3.13.3

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: GPU 0: Tesla K80
Nvidia driver version: 418.40.04
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] numpy==1.16.3
[conda] blas                      1.0                         mkl  
[conda] mkl                       2018.0.2                      1  
[conda] mkl-service               1.1.2            py36h17a0993_4  
[conda] mkl_fft                   1.0.1            py36h3010b51_0  
[conda] mkl_random                1.0.1            py36h629b387_0  
[conda] pytorch-nightly           1.2.0.dev20190621 py3.6_cuda10.0.130_cudnn7.5.1_0    pytorch
```

**Comment by vishwakftw** [comment]
Thank you for the bug report. cc: @neerajprad

**Comment by vishwakftw** [comment]
I'm not able to reproduce the failure 100%, for instance when k = 100, the p-value is 1.000 and the KL divergence is 0.000. There were many other instances like these, so I'm not sure if it were just some form of statistical inconsistency.

**Comment by LeviViana** [comment]
It seems to me that the sampling is really messed up.

![image](https://user-images.githubusercontent.com/17881241/59999709-ed144200-9662-11e9-8634-737da9af3548.png)

**Comment by LeviViana** [comment]
I did these visual distribution checks when I was implementing `torch.choice`. Here is a notebook where I made similar tests: https://gist.github.com/LeviViana/c01011a885cffc7aa434fe099b92b0df

**Comment by LeviViana** [comment]
The same thing, in another environment, using the version 1.1.0.dev20190522:

![image](https://user-images.githubusercontent.com/17881241/60034430-ef9a8a00-96aa-11e9-8fc6-e5cbb3fe48ef.png)

**Comment by fmassa** [comment]
cc @syed-ahmed maybe this is related to https://github.com/pytorch/pytorch/pull/21555 ?

**Comment by syed-ahmed** [comment]
I'm looking at this now. Let me get back to this issue in a moment.

**Comment by vishwakftw** [comment]
Closed by #22183
